Podip forwarding

greghaynes opened the pull request:
<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Related-To: #3885 



**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Switch to redesigned load balancing and connection throttler in activator which can forward requests directly to podIPs and reduces request lacency.
```

greghaynes commented:
/hold

This depends on https://github.com/knative/serving/pull/4952, probably easiest to review by looking at the third commit.

taragu commented on pkg/activator/handler/handler.go:
```diff
@@ -173,30 +113,26 @@ func (a *activationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-		configurationName := revision.Labels[serving.ConfigurationLabelKey]
-		serviceName := revision.Labels[serving.ServiceLabelKey]
-		a.reporter.ReportRequestCount(namespace, serviceName, configurationName, name, httpStatus, attempts, 1.0)
+		// Once we see a successful probe, send traffic.
```
Why is `probeEndpoint(...)` removed? Polling is done in revision backend manager's `run`, which calls [`checkDests()`](https://github.com/knative/serving/blob/63fc7bbe26fad6b5a328aaa7bcb1f48cb89800aa/pkg/activator/net/revision_backends.go#L209) to probe cluster IP and pod IPs, but polling right before sending traffic is still necessary, right?

taragu commented on pkg/activator/handler/handler.go:
```diff
@@ -173,30 +113,26 @@ func (a *activationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		proxySpan.End()
+
+		configurationName := "UNKNOWN"
+		serviceName := "UNKNOWN"
```
Do we still need to fetch revision in order to get the config and service names?

greghaynes replied:
I put this back, but IMO we really don't want to be calling in to k8s clients at all during request path. We can figure out later how we want to remove it though.

vagababov commented on pkg/activator/net/helpers.go:
```diff
@@ -0,0 +1,61 @@
+/*
```
This file should've been merged already, no?

vagababov commented:
Let me know when you merge upstream and fix the tests, I'll take a look.

greghaynes commented:
@vagababov Yep ill un-wip then. Pushed this rev up because it should deploy and be usable at this point if anyone wants to see if this fixes the issues we expect it to. I think upstream issue is waiting on CI de-flaking...

mattmoor commented on cmd/activator/main.go:
```diff
@@ -176,8 +177,20 @@ func main() {
 	params := queue.BreakerParams{QueueDepth: breakerQueueDepth, MaxConcurrency: breakerMaxConcurrency, InitialCapacity: 0}
-	throttler := activator.NewThrottler(params, endpointInformer, sksInformer.Lister(), revisionInformer.Lister(), logger)
+
+	stopCh := make(chan struct{})
```
I think you want `ctx.Done()`, which will close when we receive a signal.

mattmoor commented on pkg/activator/handler/handler.go:
```diff
@@ -98,55 +92,11 @@ func withOrigProto(or *http.Request) prober.Preparer {
 	namespace := pkghttp.LastHeaderValue(r.Header, activator.RevisionHeaderNamespace)
 	name := pkghttp.LastHeaderValue(r.Header, activator.RevisionHeaderName)
-	revID := activator.RevisionID{Namespace: namespace, Name: name}
+	//start := time.Now()
```
remove?

mattmoor commented on pkg/activator/handler/handler.go:
```diff
@@ -156,59 +106,32 @@ func (a *activationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		serviceName := revision.Labels[serving.ServiceLabelKey]
-		a.reporter.ReportRequestCount(namespace, serviceName, configurationName, name, httpStatus, attempts)
+		a.reporter.ReportRequestCount(namespace, serviceName, configurationName, name, httpStatus, 1)
+		//a.reporter.ReportResponseTime(namespace, serviceName, configurationName, name, httpStatus, time.Since(start))
```
Ah, maybe start was needed for this?

mattmoor commented on cmd/activator/main.go:
```diff
@@ -176,8 +177,20 @@ func main() {
+
+	// Start revision backends manager
+	_ = activatornet.NewRevisionBackendsManager(stopCh, revDestsCh, network.AutoTransport, revisionInformer.Lister(),
+		serviceInformer.Lister(), endpointInformer, logger)
```
If we just make this take `ctx` and use the injection stuff, then it would be like 3 parameters :)

mattmoor replied:
FWIW, I'd TODO this to get the rest baking.

mattmoor commented on pkg/activator/net/revision_backends.go:
```diff
@@ -234,21 +242,34 @@ func (rw *revisionWatcher) checkDests(dests []string) {
+		default:
+		}
+		select {
+		case <-rw.doneCh:
```
Why are these `select` statements separate?  comment?

greghaynes commented:
Un-WIP'd because I think the only things left for me to change are very small - id like to figure out if I should keep any of the commented out tests in handler_test after lunch (I think we duplicate a lot of these cases in our new throttler test).

mattmoor commented on pkg/activator/testing/roundtripper.go:
```diff
@@ -102,8 +102,10 @@ func (rt *FakeRoundTripper) popResponse(host string) *FakeResponse {
 		return resp
 	}
 
+	//fmt.Printf("Checking probe responses: %v\n", rt.ProbeResponses)
```
remove

mattmoor commented on pkg/activator/testing/roundtripper.go:
```diff
@@ -102,8 +102,10 @@ func (rt *FakeRoundTripper) popResponse(host string) *FakeResponse {
+	//fmt.Printf("Checking probe responses: %v\n", rt.ProbeResponses)
 	resp, responses := popResponseSlice(rt.ProbeResponses)
 	rt.ProbeResponses = responses
+	//fmt.Printf("Probing %q, got %v\n. Responses: %v\n", host, resp, rt.ProbeResponses)
```
remove

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -234,21 +242,34 @@ func (rw *revisionWatcher) checkDests(dests []string) {
 	for {
 		select {
+		case <-rw.doneCh:
+			// shutdown
```
This comment seems excessive.

greghaynes commented:
trying to work out these test issues before addressing comments, stand by...

greghaynes commented:
ok I think the last unit test flake is a timeout race, addressed comments and made the timeouts adjustable per test / bumped the ones which were failing.

vagababov commented:
So you want another review now? Since the last UT failure is a race :)

greghaynes commented:
@vagababov yep I think good for another review

vagababov commented on pkg/activator/handler/handler.go:
```diff
@@ -156,59 +104,32 @@ func (a *activationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			Scheme: "http",
+			Host:   dest,
 		}
+		// Once we see a successful probe, send traffic.
```
This comment no longer makes sense, since we do not _see_ any probes here.

vagababov commented on pkg/activator/handler/handler.go:
```diff
@@ -217,8 +138,8 @@ func (a *activationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
-		if err == activator.ErrActivatorOverload {
-			http.Error(w, activator.ErrActivatorOverload.Error(), http.StatusServiceUnavailable)
+		if err == activatornet.ErrActivatorOverload {
```
Since you're touching here, can I ask to return and unindent the `else`?

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -189,6 +191,24 @@ func (rw *revisionWatcher) probePodIPs(dests []string) (sets.String, error) {
 
+func (rw *revisionWatcher) sendUpdate(update *RevisionDestsUpdate) bool {
+	// In many cases we are doing blocking work while determining whether we have an update.
+	// we could have been asked to shut down and updateCh / doneCh are closed. Force a
```
```suggestion
	// We could have been asked to shut down and thus updateCh / doneCh are closed. Force a
```

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -189,6 +191,24 @@ func (rw *revisionWatcher) probePodIPs(dests []string) (sets.String, error) {
+	// In many cases we are doing blocking work while determining whether we have an update.
+	// we could have been asked to shut down and updateCh / doneCh are closed. Force a
+	// check of doneCh first in case this is the situation.
+	select {
```
I don't really see how this helps the race condition. It just makes it less probable. But the channels _can_ be closed between end of first select and start of second select. Thus still having potential of sending update on the closed channel.

greghaynes replied:
correct. That being said, in practice it seems to completely eliminate this race condition which really only shows up during test cleanup. I don't have a better idea of how to get rid of this but suggestions welcome.

greghaynes replied:
Updated per your suggestion, thanks!

vagababov commented:
You can invert the situation, where the watcher creates and owns the
channel (and returns it from new).
Now on closing doneCh, you can close the update channel as well.

On Wed, Sep 4, 2019 at 3:50 PM Gregory Haynes <notifications@github.com>
wrote:

> *@greghaynes* commented on this pull request.
> ------------------------------
>
> In pkg/activator/net/revision_backends.go
> <https://github.com/knative/serving/pull/5011#discussion_r321007305>:
>
> > @@ -189,6 +191,24 @@ func (rw *revisionWatcher) probePodIPs(dests []string) (sets.String, error) {
>  	return hs, err
>  }
>
> +func (rw *revisionWatcher) sendUpdate(update *RevisionDestsUpdate) bool {
> +	// In many cases we are doing blocking work while determining whether we have an update.
> +	// we could have been asked to shut down and updateCh / doneCh are closed. Force a
> +	// check of doneCh first in case this is the situation.
> +	select {
>
> correct. That being said, in practice it seems to completely eliminate
> this race condition which really only shows up during test cleanup. I don't
> have a better idea of how to get rid of this but suggestions welcome.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/knative/serving/pull/5011?email_source=notifications&email_token=AAF2WX2J5MOWHZE2TWMXCDTQIA3SHA5CNFSM4IIL5PG2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDWIMAQ#discussion_r321007305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAF2WX3S3YBHB5WCLNV34VTQIA3SHANCNFSM4IIL5PGQ>
> .
>

greghaynes commented:
I dont follow how we can invert it where we can still avoid this race - even if activator/cmd is responsible for closing both doneCh and updateCh it is still possible were busy probing endpoints/etc while both are closed and we race to write to updateCh?

I just noticed that were really violating some interfaces with the activator/handler test anyhow - it doesnt need to be running a revbackendsmanager at all, we can simply treat throttler as a black box which receives dests updates. This lets me remove revisionsbackendmanager entirely from the handler tests which should avoid this race.

vagababov commented:
No, what I meant, is that the update channel is owned by the updater.
you basically do this
```
select {
case _, ok <- doneCh:
  if !ok { close(updateCh); return}
default:
  updateCh <- update
}
```

Now you inject updatec hannel when you create the watcher, but instead watcher constructor can create one.

greghaynes commented:
ok this should be passing tests regularly now :crossed_fingers: and I updated to address comments.

mattmoor commented:
/hold cancel

Nope let's get it baking.

mattmoor commented:
It looks like this degrades all of the dataplane prober scenarios that involve the activator and CC.

[Here's](https://mako.dev/benchmark?benchmark_key=5142965274017792&~qc=p95000&~qc1=p95000&~qct=p95000&tseconds=172800&yscale=log&~kd=p95000&~id=p95000&~qp=p95000&~a=p95000) everything else (fairly stable):

![image](https://user-images.githubusercontent.com/2442466/64311624-a2b01b00-cf59-11e9-9df7-8e7be3d89e04.png)


[Here's](https://mako.dev/benchmark?benchmark_key=5142965274017792&~ac=p95000&~act=p95000&~ac1=p95000&tseconds=172800&yscale=log) the activator probes that specify CC:

![image](https://user-images.githubusercontent.com/2442466/64311644-b9567200-cf59-11e9-8460-e781955c3c8a.png)

greghaynes commented:
Hrm, heres a guess https://github.com/knative/serving/pull/5401

In these tests, when we increase CC are we keeping a static number of pods for the application (i.e. are we increasing total request concurrency as well?). Not sure what a good way for me to test this, but if that above PR doesnt work we'll likely need to break out pprof.

mattmoor commented:
You can see everything [here](https://github.com/knative/serving/blob/master/test/performance/dataplane-probe/dataplane-probe-setup.yaml), they are fixed scales and adequately provisioned for the 1k QPS (given their CC).

greghaynes commented:
ok, im asking because I suspect its the write locking I am fixing in this PR which would make sense if we keep a consistent pod count but update CC (which has the effect of increasing overall request concurrency and therefore lock contention here).
